src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Dialogs/DefaultWaitDialogFactory.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor.Dialogs
+{
+    [Export(typeof(WaitDialogFactory))]
+    internal class DefaultWaitDialogFactory : WaitDialogFactory
+    {
+        private readonly IVsThreadedWaitDialogFactory _waitDialogFactory;
+        private readonly JoinableTaskFactory _joinableTaskFactory;
+
+        [ImportingConstructor]
+        public DefaultWaitDialogFactory(JoinableTaskContext joinableTaskContext)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
+            _waitDialogFactory = (IVsThreadedWaitDialogFactory)Shell.Package.GetGlobalService(typeof(SVsThreadedWaitDialogFactory));
+            if (_waitDialogFactory == null)
+            {
+                throw new ArgumentNullException(nameof(_waitDialogFactory));
+            }
+
+            _joinableTaskFactory = joinableTaskContext.Factory;
+        }
+
+        // Test constructor
+        internal DefaultWaitDialogFactory(
+            JoinableTaskContext joinableTaskContext,
+            IVsThreadedWaitDialogFactory waitDialogFactory)
+        {
+
+            _waitDialogFactory = waitDialogFactory;
+            _joinableTaskFactory = joinableTaskContext.Factory;
+        }
+
+        public override WaitDialogResult<TResult> TryCreateWaitDialog<TResult>(string title, string message, Func<WaitDialogContext, Task<TResult>> onWaitAsync)
+        {
+            Debug.Assert(_joinableTaskFactory.Context.IsOnMainThread);
```

Comment:
If it's not a main thread can we proceed?

Reply:
So i'm not entirely sure about the thread restrictions for the underlying wait dialog infrastructure pieces but I got some warnings from VS analyzers so added this just to be careful